MatrixSymbol first arg is a Symbol (and other matrix expressions cleanup)

asmeurer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

The main change here is to make it so that `MatrixSymbol.args[0]` is a Symbol instead of a string. `MatrixSymbol.name` is still a string. (you can still create it with a string, it just gets converted to a Symbol in the .args automatically)

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

I still need to do an issue tracker sweep to see what kinds of issues are affected. I know at least the printing issue from https://github.com/sympy/sympy/issues/15120 is fixed here. 

#### Brief description of what is fixed or changed

See the commit messages for the minor cleanups.

#### Other comments

This has been a huge issue ever since the creation of MatrixSymbol. The problem is that non-Basic args break any code that assumes args are Basic. Most likely there is a lot of code that special-cases MatrixSymbol that can be removed here. I still need to see if any tests fail first. 

This is a backwards incompatible break. I don't think there's any way to do a deprecation.  I've left the type of `MatrixSymbol.name` intact to keep as much compatibility as possible. 

## TODO: 

- [ ] Add tests for
  - [ ] #15120
  - [ ] #6249
  - [x] https://github.com/sympy/sympy/issues/6913 (works in master but make sure it is tested)
  - [x] https://github.com/sympy/sympy/issues/2750 (works in master)
  - [ ] https://github.com/sympy/sympy/issues/7842 (works in master)
  - [ ] https://github.com/sympy/sympy/issues/9817 (str related issue works in master)
  - [ ] https://github.com/sympy/sympy/issues/11948 (not sure the exact issue but it seems to work in master)
  - [ ] https://github.com/sympy/sympy/pull/13453 (the recursion error is gone, but there's a fixed printing error from master)
  - [ ] Check if https://github.com/sympy/sympy/pull/14424 is still necessary (c.f. https://github.com/sympy/sympy/issues/14421, which is broken in master and fixed here)
  - [ ] https://github.com/sympy/sympy/issues/15050 (no longer errors but doesn't give a solution)
  - [ ] Check for other issues
- [x] See if removing the `isinstance(..., string_types)` fails any tests. https://github.com/sympy/sympy/pull/15121#issuecomment-423315973
  - [x] Check for other special casing of string args
- [x] Fix `exp(A)*exp(B)` https://github.com/sympy/sympy/pull/15121#issuecomment-442182341
- [x] XFAIL incorrect matrix derivative test https://github.com/sympy/sympy/pull/15121#discussion_r242716398
#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. If there is no release notes entry for this PR,
write "NO ENTRY". The bot will check your release notes automatically to see
if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * `count_ops` no longer handles objects with strings in their `.args` (all elements of `.args` for SymPy objects should be `Basic` subclasses).
* matrices
  * `MatrixSymbol.args[0]` is now a `Symbol` instead of a string. `MatrixSymbol.name` is still a string.
* printing
  * srepr now correctly prints subclasses of `Add` and `Mul`.
<!-- END RELEASE NOTES -->

asmeurer committed:
Fix undefined name in MatrixExpression._eval_derivative

The intended result of this code, replacing a trace with a summation, doesn't
seem to happen, but at least it no longer results in a NameError.

asmeurer committed:
Use _sympify instead of sympify in matexpr.py

This makes it so that string inputs do not work. This does not (yet) change
the string argument of MatrixSymbol.

asmeurer committed:
Store the first argument of MatrixSymbol as a Symbol

MatrixSymbol.name is still a string.

This is a backwards incompatible change, but one that has needed to be changed
for some time. The issue is that various places in SymPy assume that all args
are SymPy types, and hence fail if they are not. This is exacerbated by
functions that rebuild expressions after sympifying the args. It was possible,
for instance, with something like

A = MatrixSymbol('A', n, n)
B = MatrixSymbol('B', n, n)
factor(A*B)

to create an expression that would fail in the printers, because it replaced
MatrixSymbol('A', n, n) with MatrixSymbol(Symbol('A'), n, n). This is because
factor (actually factor_nc) does effectively

expr = sympify(expr)
return expr.func([factor_nc(i) for i in expr.args])

I have not yet gone through the issue tracker to try to find things that this
might fix, nor have I run the test suite beyond sympy/matrices, which had no
failures. The only test that I've added thus far is to un-XFAIL the test_args
test for MatrixSymbol.

asmeurer committed:
Fix some test failures from MatrixSymbol.args[0] being Symbol

Upabjojr commented on sympy/matrices/expressions/tests/test_derivatives.py:
```diff
@@ -86,3 +86,8 @@ def test_matrix_derivative_with_inverse():
+def test_trace():
+    # This should probably simplify more
+    expr = Trace(A)*A
+    assert expr.diff(A) == Derivative(expr, A)
```
What do you mean?

asmeurer replied:
Actually never mind. It is used, but only internally. But the other question remains, should `diff(Trace(A)*A, A)` simplify? I suppose the answer is probably no, given that diff(A, A) doesn't simplify. 

Also, unrelated, but `diff(Trace(A), A)` gives `1`, but the cookbook says it should be `I`.

smichr commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -596,12 +596,14 @@ class MatrixElement(Expr):
-        name = sympify(name)
+        if isinstance(name, string_types):
+            name = Symbol(name)
+        name = _sympify(name)
```
Should that `name = _sy...` be
```python
elif not isinstance(name, Symbol):
    raise TypeError('name must be string of Symbol')
```

smichr commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -658,7 +660,9 @@ class MatrixSymbol(MatrixExpr):
-        n, m = sympify(n), sympify(m)
+        n, m = _sympify(n), _sympify(m)
+        if isinstance(name, string_types):
+            name = Symbol(name)
```
elif could be added here, too?

smichr commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -596,12 +596,14 @@ class MatrixElement(Expr):
             if n.is_Integer and m.is_Integer:
                 return name[n, m]
-        name = sympify(name)
+        if isinstance(name, string_types):
```
and this could be `elif` since it's not a MatrixBase.

smichr committed:
Merge branch 'master' into matexpr-cleanup

asmeurer commented:
All the simplify functions (https://github.com/sympy/sympy/issues/6249) now work.

asmeurer commented:
There are some lines in the core like

https://github.com/sympy/sympy/blob/3d0e2ec4dcb7653f7fac039cc585ae81dc7251c2/sympy/core/function.py#L2632-L2634

https://github.com/sympy/sympy/blob/3d0e2ec4dcb7653f7fac039cc585ae81dc7251c2/sympy/core/function.py#L2723-L2725

which were definitely put there for MatrixSymbol. But there may be other things as well. I can try removing them and seeing what tests fail.

asmeurer committed:
Test commenting out test against str args in the core

With MatrixSymbol args no longer str, it may no longer be necessary. This is
just a test commit to see if any tests fail on Travis.

asmeurer commented:
When searching for issues for this, I also found many issues like https://github.com/sympy/sympy/issues/11521 (`AttributeError: 'MatrixSymbol' object has no attribute '_diff_wrt'`) that are now fixed. Someone should go through and make sure they are tested and close them.

asmeurer commented:
My TODO list above is I believe extensive. What remains to be done here is to go through each issue and make sure it is tested here, and close it once this is merged.

asmeurer commented:
Tests pass without those checks in count_ops. Should we remove them?

asmeurer commented:
It seems there is a bug here:

```pytb
>>> A = MatrixSymbol('A', 2, 2)
>>> B = MatrixSymbol('B', 2, 2)
>>> exp(A)*exp(B)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "./sympy/core/decorators.py", line 132, in binary_op_wrapper
    return func(self, other)
  File "./sympy/core/expr.py", line 140, in __mul__
    return Mul(self, other)
  File "./sympy/core/operations.py", line 45, in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
  File "./sympy/core/mul.py", line 354, in flatten
    o12 = b1 ** new_exp
  File "./sympy/core/expr.py", line 154, in __pow__
    return self._pow(other)
  File "./sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "./sympy/core/decorators.py", line 131, in binary_op_wrapper
    return f(self)
  File "./sympy/matrices/expressions/matexpr.py", line 24, in __sympifyit_wrapper
    return func(a, b)
  File "./sympy/core/decorators.py", line 132, in binary_op_wrapper
    return func(self, other)
  File "./sympy/matrices/expressions/matexpr.py", line 145, in __rpow__
    raise NotImplementedError("Matrix Power not defined")
NotImplementedError: Matrix Power not defined
```

jksuom commented:
File "./sympy/core/mul.py", line 354, in flatten
       o12 = b1 ** new_exp

This tries to compute `E**(A + B)` which is wrong. On the previous line, it is checked that `not new_exp.is_Add`, but it does not help as `A + B` is of type `MatAdd`.

jksuom commented:
`exp(A)*exp(B)` should probably not be handled by core. Maybe `exp(A)` should be a matrix.

asmeurer commented:
This is related to https://github.com/sympy/sympy/issues/15551

asmeurer committed:
Merge branch 'master' into matexpr-cleanup

asmeurer commented:
It seems the merge with master fixed the problem.

asmeurer committed:
Add a test for exp(A)*exp(B) for matrix expressions

asmeurer committed:
Merge branch 'master' into matexpr-cleanup

asmeurer commented on sympy/matrices/expressions/tests/test_derivatives.py:
```diff
@@ -101,6 +101,9 @@ def test_matrix_derivative_vectors_and_scalars():
 def test_matrix_derivatives_of_traces():
 
+    expr = Trace(A)*A
+    assert expr.diff(A) == Trace(A)*Identity(k) + A.T
```
@Upabjojr can you confirm if this is correct?

Upabjojr replied:
```
Trace(A)*A) ===> A_{ii} A_{ij}
```
now
```
diff(A_ii A_jk, A_mn) = d_mi d_ni A_jk + A_ii d_mj d_nk = d_mn A_jk + A_ii d_mj d_nk
```
There is a problem with this result, as it should probably be a higher-rank array. MatrixCalculus.org is indeed unable to compute it.

Upabjojr replied:
I'd say the result is wrong, it's omitting a higher dimensional identity.

Upabjojr replied:
https://github.com/sympy/sympy/issues/15667

asmeurer committed:
Add an additional test that matrix exponentials don't commute

asmeurer committed:
Remove print statement from a test

asmeurer committed:
Add a test that MatrixSymbol(1, 2, 'A') raises an exception

Fixes #15643.

asmeurer committed:
Remove commented out checks for string args in count_ops()

These are no longer necessary now that MatrixSymbol doesn't have non-Basic
args.

asmeurer commented:
Some of the simplification functions that worked before don't work any more, as a result of https://github.com/sympy/sympy/pull/15373. I am digging into the reasons. It looks like some code ends up creating Add or Mul instead of MatAdd or MatMul on matrix expressions.

RituRajSingh878 commented:
@asmeurer  I would like to work on these tasks.I think here we have to add test.

asmeurer committed:
Use f.func instead of f.func._from-args in cancel

Otherwise it doesn't do the right thing with matrix expressions.

asmeurer commented:
After https://github.com/sympy/sympy/pull/15948, almost all simplifying functions will raise NotImplementedError. They will need to be adapted manually to work with matrix expressions.

asmeurer commented:
Should we try to get this in the 1.4 release? It would be nice to have it, but it also isn't critical.

asmeurer commented:
Another place that handles non-Basic args https://github.com/sympy/sympy/blob/f499cbbc188d777be2998b01e48a3d8b64f1ea10/sympy/core/exprtools.py#L1093

cbm755 commented:
I ran my downstream Octave project test set on this without any additional failures, so I guess that's a partial +1 for merging into 1.4.  That is, it sounds reasonable, doesn't break anything for me, but I haven't carefully reviewed.

asmeurer committed:
Merge branch '1.4' into matexpr-cleanup

asmeurer committed:
Make srepr work correctly on Add and Mul subclasses

Fixes #16430.

asmeurer committed:
Add comments that MatAdd and MatMul probably shouldn't be subclassing from Add and Mul

There are some test failures right now if you remove them, at least in the
printers because they are using the Add and Mul printers, possibly other
places as well. For now we can leave it be but leave a comment that we might
want to remove them if they cause problems.

asmeurer commented:
Any objections to merging this into the 1.4 release branch?

smichr commented:
> all elements of .args for SymPy objects should be Basic subclasses

This is not enforced by Matrices, but they are not a Basic instance. I was surpriced that `laplace_transform` send back a matrix with lists as entries. In the work I have done recently with matrices I have tried to be as lenient as possible for Matrix contents.

BTW, I'm +1 to committing this. I see you have several TODOs so did not commit this now. I'm not sure if you want to finish those before this goes in or not.

asmeurer commented:
I'll double check them but a lot of them were things that I was going to add tests for but they now don't work after #15948.
